### DISCUSSION TITLE
otbr-agent receives the packet but does not forward it to the application through wpan0 interface.

### DISCUSSION THREAD

**Discussion by jinpeng1989** [openthread/openthread#9113] [Q&A] [answered]
**The A device send firmware to the B device via CoAP. After the new firmware is applied, the B link to the thread network again. A send CoAP Get message to read the version number of B, B reply CoAP 2.05 message, A replies ACK, but A does not pass the message to the application through wpan0 interface. In short, when device B re-link to the network after firmware update, the application in A device cannot receive data packets from device B. This anomaly usually resolves itself in a few hours.

The issue can be reproduced. Can you help diagnose it? 

## Network topology
![image](https://github.com/openthread/openthread/assets/97089879/aff6acac-25c6-4f18-ae99-51e4a428e86c)

## KNBA5F000026 (A device)
* RCP SoC: [Silicon Labs EFR32MG21](https://www.silabs.com/wireless/zigbee/efr32mg21-series-2-socs)
* Thread 1.2 Specification (per compile flag)
* Gecko SDK (GSDK) Version [v4.2.0.0](https://github.com/SiliconLabs/gecko_sdk/releases/tag/v4.2.0)
    * OpenThread [91fa1f4](https://github.com/openthread/openthread/tree/91fa1f455a1d368dff9d712b38561bd08688e1f5) (2022-11-07)
    * OpenThread Border Router [d910392](https://github.com/openthread/ot-br-posix/tree/d9103922af71d5b0a46dd7f224fd7bc2244ca1b4) (2022-11-08)

## TME (B device)
* Wireless SoC: [Silicon Labs EFR32MG21](https://www.silabs.com/wireless/zigbee/efr32mg21-series-2-socs)
* Thread 1.3 Specification
* OpenThread [8f92d2dc8](https://github.com/openthread/openthread/tree/8f92d2dc8) (2022-06-02)

The marked packets are those captured by sniffer, but not by wpan0.
![image](https://github.com/openthread/openthread/assets/97089879/39216ca2-4bde-41e1-96fc-738441ae8017)
![image](https://github.com/openthread/openthread/assets/97089879/01b10cf2-4136-48c8-83a5-c4c9a67f3e79)

**Comment by jinpeng1989** [comment]
[no_forwarding.zip](https://github.com/openthread/openthread/files/11612107/no_forwarding.zip)
If device B simply restarts and re-Link to the network, this problem does not occur. I compared the data of the two captured packets, which were restarted with updated firmware and restarted without updated firmware, and found no difference. So I wonder why the results are different.

**Reply by jwhui** [reply, marked as answer]
Looking at the sniffer trace, it appears that the MAC frame counter is not being restored properly.

- Frame#56735: MAC Frame Counter = 2202
- Frame#56740: MLE Link Request (indicates device reattaching after reset)
- Frame#56749: MAC Frame Counter = 2008

Because the MAC Frame Counter is going backwards, the receiver is reject the frames.

OpenThread stores the MAC Frame Counter here:
https://github.com/openthread/openthread/blob/8f5d4376c6b5c681ca3c426feb477825bc7dbfad/src/core/thread/mle.cpp#L510

OpenThread reads the MAC Frame Counter after reset here:
https://github.com/openthread/openthread/blob/8f5d4376c6b5c681ca3c426feb477825bc7dbfad/src/core/thread/mle.cpp#L376

Can you validate that the platform is properly reading/writing non-volatile settings information?

**Reply by jinpeng1989** [reply]
Thank you for your prompt response and support. 
I was not involved in the development of B device, so I am not in a position to help you to validate the platform API.
From what I understand, the defect has to do with the storage mechanism of device B. The networInfo is not stored directly into non-volatile memory, but is cached into RAM and stored at 2-minute intervals. 
![image](https://github.com/openthread/openthread/assets/97089879/39b1190b-984a-43d9-bd57-761953199dc7)

**Reply by jwhui** [reply]
> From what I understand, the defect has to do with the storage mechanism of device B. The networInfo is not stored directly into non-volatile memory, but is cached into RAM and stored at 2-minute intervals.

This explains the non-conformant behavior. Performing SW update will cause the Frame Counters to increment more quickly, making it more likely for the wrong Frame Counter to be used upon reset. Vendor of Device B probably needs to be made aware of the issue and have it fixed.

**Comment by jinpeng1989** [comment]
[no_forwarding_syslog.log](https://github.com/openthread/openthread/files/11614747/no_forwarding_syslog.log)

During the period from 01:50:16 to 03:27:36, there is the MeshForwarder record of device B as the target address in syslog, but no Meshforwarder record as the source address. According to the sniffer packets, device A's RCP did receive the message from device B and respond with an ACK. 
![image](https://github.com/openthread/openthread/assets/97089879/72ae1eae-77b9-4aa9-b2f3-27261f9ad151)

**Comment by jwhui** [comment]
@jinpeng1989 , can you provide the actual pcap file for analysis in Wireshark?

**Reply by jinpeng1989** [reply]
Please see the attachment above. The file name is no_forwarding.zip. It contains packet capture data of sniffer and wpan0 interface. 
![image](https://github.com/openthread/openthread/assets/97089879/07e1782a-7b70-4e69-8d88-60d1b5d12418)
